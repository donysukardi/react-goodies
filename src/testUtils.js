const propsToDataAttrs = props =>
  Object.entries(props).reduce((acc, [key, value]) => {
    acc[`data-${String(key).toLowerCase()}`] = value
    return acc
  }, {})

export {propsToDataAttrs}
