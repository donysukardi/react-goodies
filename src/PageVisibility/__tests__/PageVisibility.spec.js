/* eslint-disable global-require */
import React from 'react'
import { render } from 'react-testing-library'
import { propsToDataAttrs } from '../../testUtils'

const Target = props => (
  <div data-testid="targetContainer" {...propsToDataAttrs(props)} />
)

const renderTarget = props => <Target {...props} />

describe('PageVisibility', () => {
  describe('Page Visibility API is supported', () => {
    let PageVisibility = null
    let origVisibilityState = null

    beforeAll(() => {
      origVisibilityState = global.document.visibilityState
    })

    beforeEach(() => {
      jest.resetModules()

      Object.defineProperty(global.document, 'visibilityState', {
        get: () => 'visible',
        configurable: true,
      })
      PageVisibility = require('../PageVisibility').default
    })

    afterAll(() => {
      Object.defineProperty(global.document, 'visibilityState', {
        get: () => origVisibilityState,
      })
    })

    it('should set initial state', () => {
      const { getByTestId } = render(
        <PageVisibility>{renderTarget}</PageVisibility>,
      )

      expect(getByTestId('targetContainer')).toMatchSnapshot()
    })

    it('should handle page visibility state change', () => {
      const { getByTestId } = render(
        <PageVisibility>{renderTarget}</PageVisibility>,
      )

      Object.defineProperty(global.document, 'visibilityState', {
        get: () => 'hidden',
        configurable: true,
      })
      global.document.dispatchEvent(new CustomEvent('visibilitychange'))

      expect(getByTestId('targetContainer')).toMatchSnapshot()
    })

    it('should remove event listener on unrender', () => {
      const mockVisibility = jest.fn(() => 'hidden')
      const { unmount } = render(
        <PageVisibility>{renderTarget}</PageVisibility>,
      )

      unmount()

      Object.defineProperty(global.document, 'visibilityState', {
        get: mockVisibility,
        configurable: true,
      })
      global.document.dispatchEvent(new CustomEvent('visibilitychange'))

      expect(mockVisibility.mock.calls).toMatchSnapshot()
    })
  })
})
