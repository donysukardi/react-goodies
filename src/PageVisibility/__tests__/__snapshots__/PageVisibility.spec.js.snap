// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageVisibility Page Visibility API is supported should handle page visibility state change 1`] = `
<div
  data-ishidden="true"
  data-isprerendered="false"
  data-isunloaded="false"
  data-isvisible="false"
  data-testid="targetContainer"
/>
`;

exports[`PageVisibility Page Visibility API is supported should remove event listener on unrender 1`] = `
Array [
  Array [],
  Array [],
]
`;

exports[`PageVisibility Page Visibility API is supported should set initial state 1`] = `
<div
  data-ishidden="false"
  data-isprerendered="false"
  data-isunloaded="false"
  data-isvisible="true"
  data-testid="targetContainer"
/>
`;
